packages/ckeditor5-widget/src/widgettypearound/widgettypearound.js

```diff
@@ -0,0 +1,250 @@
+		this._enableInsertingParagraphsOnButtonClick();
+	}
+
+	/**
+	 * Inserts a new paragraph next to a widget element with the selection anchored in it.
+	 *
+	 * **Note**: This method is heavily user-oriented and will both focus the editing view and scroll
+	 * the viewport to the selection in the inserted paragraph.
+	 *
+	 * @protected
+	 * @param {module:engine/view/element~Element} widgetViewElement The view widget element next to which a paragraph is inserted.
+	 * @param {'before'|'after'} position The position where the paragraph is inserted. Either `'before'` or `'after'` the widget.
+	 */
+	_insertParagraph( widgetViewElement, position ) {
+		const editor = this.editor;
+		const editingView = editor.editing.view;
+		let viewPosition;
+
+		if ( position === 'before' ) {
+			viewPosition = editingView.createPositionBefore( widgetViewElement );
+		} else {
+			viewPosition = editingView.createPositionAfter( widgetViewElement );
+		}
+
+		const modelPosition = editor.editing.mapper.toModelPosition( viewPosition );
+
+		editor.model.change( writer => {
+			const paragraph = writer.createElement( 'paragraph' );
+
+			writer.insert( paragraph, modelPosition );
+			writer.setSelection( paragraph, 0 );
+		} );
+
+		editingView.focus();
+		editingView.scrollToTheSelection();
+	}
+
+	/**
+	 * Creates a listener in the editing conversion pipeline that injects the type around
+	 * UI into every single widget instance created in the editor.
+	 *
+	 * The UI is delivered as a {@link module:engine/view/uielement~UIElement}
+	 * wrapper which renders DOM buttons that users can use to insert paragraphs.
+	 *
+	 * @private
+	 */
+	_enableTypeAroundUIInjection() {
+		const editor = this.editor;
+		const schema = editor.model.schema;
+		const t = editor.locale.t;
+		const buttonTitles = {
+			before: t( 'Insert paragraph before widget' ),
+			after: t( 'Insert paragraph after widget' )
+		};
+
+		editor.editing.downcastDispatcher.on( 'insert', ( evt, data, conversionApi ) => {
+			const viewElement = conversionApi.mapper.toViewElement( data.item );
+
+			// Filter out non-widgets and inline widgets.
+			if ( isTypeAroundWidget( viewElement, data.item, schema ) ) {
+				injectUIIntoWidget( editor.editing.view, buttonTitles, viewElement );
+			}
+		}, { priority: 'low' } );
+	}
+
+	/**
+	 * Registers an editing view post-fixer which checks all block widgets in the content
+	 * and adds CSS classes to these which should have the typing around (UI) enabled
+	 * and visible for the users.
+	 *
+	 * @private
+	 */
+	_enableDetectionOfTypeAroundWidgets() {
+		const editor = this.editor;
+		const schema = editor.model.schema;
+		const editingView = editor.editing.view;
+
+		function positionToWidgetCssClass( position ) {
+			return `ck-widget_can-type-around_${ position }`;
+		}
+
+		editingView.document.registerPostFixer( writer => {
+			// Find all view elements in the editing root.
+			[ ...editingView.createRangeIn( editingView.document.getRoot() ) ]
```

Comment:
This would be really heavy due to TreeWalker. I'd replace this search for widgets enhanced with the converter defined in `_enableTypeAroundUIInjection` with a set of these widgets. The aforementioned converter would add new elements to that set and the postfixer here could remove them once they [are not attached any more](https://ckeditor.com/docs/ckeditor5/latest/api/module_engine_view_element-Element.html#function-isAttached).